lib/mayaUsd/nodes/proxyShapeListenerBase.cpp

```diff
@@ -0,0 +1,227 @@
+void* MayaUsdProxyShapeListenerBase::creator() { return new MayaUsdProxyShapeListenerBase(); }
+
+/* static */
+MStatus MayaUsdProxyShapeListenerBase::initialize()
+{
+    MStatus retValue = MS::kSuccess;
+
+    //
+    // create attr factories
+    //
+    MFnNumericAttribute numericAttrFn;
+
+    stageCacheIdAttr
+        = numericAttrFn.create("stageCacheId", "stcid", MFnNumericData::kInt, -1, &retValue);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    numericAttrFn.setStorable(false);
+    numericAttrFn.setConnectable(false);
+    numericAttrFn.setReadable(false);
+    retValue = addAttribute(stageCacheIdAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    outStageCacheIdAttr
+        = numericAttrFn.create("outStageCacheId", "ostcid", MFnNumericData::kInt, -1, &retValue);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    numericAttrFn.setStorable(false);
+    numericAttrFn.setConnectable(true);
+    numericAttrFn.setWritable(false);
+    retValue = addAttribute(outStageCacheIdAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    //
+    // Smart signaling attributes:
+    //
+    updateCounterAttr
+        = numericAttrFn.create("updateId", "upid", MFnNumericData::kInt64, 0, &retValue);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    numericAttrFn.setStorable(false);
+    numericAttrFn.setHidden(true);
+    numericAttrFn.setIndeterminant(true);
+    numericAttrFn.setWritable(false);
+    numericAttrFn.setCached(false);
+    numericAttrFn.setAffectsAppearance(false);
+    retValue = addAttribute(updateCounterAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    resyncCounterAttr
+        = numericAttrFn.create("resyncId", "rsid", MFnNumericData::kInt64, 0, &retValue);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    numericAttrFn.setStorable(false);
+    numericAttrFn.setHidden(true);
+    numericAttrFn.setIndeterminant(true);
+    numericAttrFn.setWritable(false);
+    numericAttrFn.setCached(false);
+    numericAttrFn.setAffectsAppearance(false);
+    retValue = addAttribute(resyncCounterAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    //
+    // add attribute dependencies
+    //
+    retValue = attributeAffects(stageCacheIdAttr, outStageCacheIdAttr);
+    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+    return retValue;
+}
+
+void MayaUsdProxyShapeListenerBase::postConstructor() { setExistWithoutInConnections(false); }
+
+MStatus MayaUsdProxyShapeListenerBase::compute(const MPlug& plug, MDataBlock& dataBlock)
+{
+    if (plug == outStageCacheIdAttr) {
+        MStatus retValue = MS::kSuccess;
+
+        MDataHandle inHandle = dataBlock.inputValue(stageCacheIdAttr, &retValue);
+        CHECK_MSTATUS_AND_RETURN_IT(retValue);
+
+        int cacheIdNum = inHandle.asInt();
+
+        if (cacheIdNum != _lastKnownStageCacheId) {
+            const auto cacheId = UsdStageCache::Id::FromLongInt(cacheIdNum);
+            const auto stageCached
+                = cacheId.IsValid() && UsdUtilsStageCache::Get().Contains(cacheId);
```

Comment:
Don't we have multiple stage cache in MayaUSD, for shared/unshared stages, and payload? See stageCache.h.

Reply:
I am aware of these caches now. But it looks like our clients prefer a cache accessible directly via USD APIs. This is also the cache used by the proxyShape for its `outStageCacheId` output, so I am not introducing anything new here.